### PR TITLE
Add an inline attribute on functions

### PR DESCRIPTION
**Pull request by chambart**
This PR propose to add an attribute to enforce and ensure that a function is inlinable or inlined.

The added attributes are:
### `inline` or `ocaml.inline` to annotate function that should always be inlined

It can either be placed on the function expression or on the let binding:

``` ocaml
let f = (fun x -> x) [@inline]
let g x = x [@@inline]
```

If any direct call is made on `f` and `f` is inlinable, it will be inlined.

It also accept `force` or `never` as payload. `never` prevents the function from being inlined.
`force` is equivalent to no payload.
### `inlined` or `ocaml.inlined` to annotate a call site

Annotating a call site with inlined premit to ensure that a call is effectively inlined. A warning
is produced otherwise.

``` ocaml
let f x = x
let g x = (f [@inlined]) x
```

It also accept the same payload. `never` prevents the call from being inlined.
#### Also

As a side effect of introducing checks for the admissible position for those attributes,
I also added checks for `[@tailcall]` (and `[ocaml.tailcall]` is accepted)

### CODE REVIEW

**Comment by trefis** [comment]
I haven't looked at the patches, but I have a few comments/questions regarding the description : 

## 

> ``` ocaml
> let f x = x + 1 [@inline] (* on the 1 expression *)
> ```
> 
> Should be a syntax error or a warning.

This is not a syntax error.

## 

> Should the attribute be an extension node

No. A lengthy discussion already took place on PR #111 , I think @gasche made a pretty good point in [this comment](https://github.com/ocaml/ocaml/pull/111#issuecomment-61231736).

## 

> Notice that we don't want to annotate the let expression itself, because it could contain multiple function declarations:
> 
> ``` ocaml
> let f x = x + 1
> and g x = x + 2
>  [@@inline]
> ```

I'm not sure what the problem here is. If I read the parser correctly, each (toplevel) value binding can have its own (post item) attributes.
Or is the problem that when you have multiple function declarations, you want/have to inline them all and they should all have the "inline" attribute?

Note: I said I didn't want the attribute to be an extension node, i.e. I don't like the `let f = [%inline fun ...]`. But I actually quite like the `let%inline f x = ...`, in particular because for expression we can't just add `[@inline]`/`[@@inline]` after a binding, and adding the attribute on a pattern as proposed feels dirty. But have `let%inline f x = ...` be rewritten to `let f = (fun x -> ...) [@inline]` seems "OK".

**Comment by trefis** [comment]
For the check regarding misuse of the attribute, how about something like https://github.com/trefis/ocaml/commit/34ebb7f8dc7b31813b82b92c105b7fc976314ad0 ?

File: asmcomp/closure.ml [hunk 1]
```diff
@@ -1080,12 +1083,16 @@ and close_functions fenv cenv fun_defs =
...
-               split_default_wrapper id kind params body
+           | (id, Lfunction{kind; params; body; attr}) ->
+               split_default_wrapper id kind params body attr
            | _ -> assert false
          )
          fun_defs)
   in
+  let inline_attribute = match fun_defs with
+    | [_, Lfunction{kind; params; body; attr = { inline }}] -> inline
+    | _ -> false (* recursive functions can't be inlined *)
```
Comment (yallop) [inline asmcomp/closure.ml]: It'd be better not to silently ignore the `@inline` attribute for recursive functions.  In any case, doesn't the compiler sometimes inline recursive functions?

**Comment by chambart** [comment]
Effectively I missed something about the attribute on let. So if we want an annotation on the let I would propose

``` ocaml
let f x = x + 1
  [@@inline]
and g y = y + 2
and h z = z
  [@@inline]
```

But following @gasche argument I suppose that we don't want 2 places for this annotation, hence if we keep only one this should go on the function expression. Otherwise we would force adding a let binding to add the annotation.

Thank you for the warning patch, I will pick it.

File: asmcomp/closure.ml (hunk 1 above)
Comment (chambart) [inline asmcomp/closure.ml]: That's the quick and dirty way for closure. It will gets proper treatment on flambda.

**Comment by Drup** [comment]
As you already know, my vote goes for `let[@inline] f x = x+1`
Related mantis ticket: http://caml.inria.fr/mantis/view.php?id=6586

Edit: as @trefis said, attributes on top level let is per-binding, so the and is not an issue.

**Comment by Drup** [comment]
@chambart To be consistent with the other attributes exposed by ocaml, the function `has_inline_attribute` should also accept the form "ocaml.inline".

**Comment by chambart** [comment]
Should I fix the [@tailcall] attribute also ?

**Comment by lpw25** [comment]
I think that the annotation should go on the `let` binding itself, because annotating expressions with `[@inline]` should be reserved for requesting that a particular call be inlined.

**Comment by chambart** [comment]
@lpw25, I would rather use a different annotation for the application, probably `[@inlined]`

**Comment by lpw25** [comment]
> @lpw25, I would rather use a different annotation for the application, probably [@inlined]

Fair enough.

> But following @gasche argument I suppose that we don't want 2 places for this annotation, hence if we keep only one this should go on the function expression.

Then I would argue that it should be allowed on both let expressions and function expressions since they are both forms that introduce functions. (This is certainly the case for let expressions which have parameters -- for let expressions without parameters I suppose that it is less clear).

**Comment by damiendoligez** [comment]
> Then I would argue that it should be allowed on both let expressions and function expressions since they are both forms that introduce functions.

Indeed, Gabriel's argument doesn't apply here because we already have two syntaxes for defining functions. It feels wrong to provide the attribute for one of them but not the other.

**Comment by chambart** [comment]
Updated to introduce `[@inlined]` on call site and `let f x = x [@@inline]` on let bindings

**Comment by lpw25** [comment]
Perhaps `always` would be better than `force` -- since it is dual to `never`.

**Comment by chambart** [comment]
Effectively.
Renaming done.

File: utils/warnings.mli [hunk 2]
```diff
@@ -63,6 +63,9 @@ type t =
   | Eliminated_optional_arguments of string list (* 48 *)
   | No_cmi_file of string                   (* 49 *)
   | Expect_tailcall                         (* 50 *)
+  | Missplaced_attribute of string          (* 51 *)
```
Comment (damiendoligez) [inline utils/warnings.mli]: spelling: should be `Misplaced_attribute`

Comment (chambart) [inline utils/warnings.mli]: Thanks for the catch

**Comment by lpw25** [comment]
I see that attributes are extracted during translcore. I think perhaps it would be better if they were extracted during typecore. For a start, this would produce the warning for an incorrect paylod during typechecking, which would seem a better place for it.

It would also be good to allow the `[@@inline]` attribute on interfaces. Such as:

``` ocaml
val f: int -> int [@@inline]
```

but perhaps this should be saved for a later patch.

**Comment by chambart** [comment]
What kind of behaviour would you expect from `[@@inline]` on an interface ? Failing if the function is not annotated with inline in the source file or propagate the annotation to the implementation declaration ?

**Comment by lpw25** [comment]
I would expect uses of the function in other modules to warn if they are not inlined. In particular, I would expect a warning if .cmx files were unavailable for this module.

**Comment by mshinwell** [comment]
I haven't thought about the details as to the exact positioning of these attributes, but I think they should be considered essential in conjunction with Flambda.

File: bytecomp/translcore.ml [hunk 3]
```diff
@@ -646,10 +649,120 @@ let rec cut n l =
 
 let try_ids = Hashtbl.create 8
 
-let has_tailcall_attribute e =
-  List.exists (fun ({txt},_) -> txt="tailcall") e.exp_attributes
+let is_inline_attribute = function
```
Comment (gasche) [inline bytecomp/translcore.ml]: I think the fact that we have 100 lines of code dedicated to attribute handling here is the sign that something is wrong. Could we factorize the logic (there are good patterns implemented there) and move this to a utility module?

**Comment by gasche** [comment]
The code seems ok, except that we should not have a big blurb dedicated to attribute handling in the middle of translcore (this is bound to be useful in other places in the compiler anyway). I would be interested in eventually merging the patch, if only for its very useful and descriptive tests which also cover `@tailcall`.

**Comment by mshinwell** [comment]
I have updated this patch to split the attribute handling into a separate file as requested by gasche.  I would like to merge this pull request soon unless there are any further objections.

**Comment by gasche** [comment]
Your commits contain some remaining conflict stuff (`git grep "<<<<<"` should work) related to warning numbers. Otherwise I like the new global code organization. (I haven't reviewed the details of the patch but if you and Pierre agree on it you should definitely go ahead.)

**Comment by mshinwell** [comment]
Whoops, I'm not sure how that happened.  Fixed, re-tested and pushed to trunk.

**Comment by chambart** [comment]
Merged in svn commit 16530

**Pull request closed without merging**